src/lib/ast.rs

```diff
@@ -0,0 +1,623 @@
+    pub fn render_dotgraph<W: Write>(&self, buffer: &mut W) -> Result<(), io::Error> {
+        dot::render(self, buffer)
+    }
+}
+
+impl<'a> dot::Labeller<'a, NodeId, EdgeId> for Arena<String, String> {
+    // The graph and node ids here must be non-empty strings made up of
+    // alphanumeric or underscore characters, not beginning with a digit (i.e.
+    // the regular expression [a-zA-Z_][a-zA-Z_0-9]*). This means we cannot
+    // label nodes with arithmetic operators, etc.
+
+    fn graph_id(&self) -> dot::Id {
+        dot::Id::new("AST").unwrap()
+    }
+
+    fn node_id(&self, id: &NodeId) -> dot::Id {
+        // Node ids must be unique.
+        dot::Id::new(format!("N{}", id.index)).unwrap()
+    }
+
+    fn node_label(&self, id: &NodeId) -> dot::LabelText {
+        let mut label = String::new();
+        label.push_str(self.nodes[id.index].ty.as_str());
+        // Only add node.data to the label if it is alphanumeric.
+        let mut is_data_alphanum = true;
+        if self.nodes[id.index].data.len() > 0 {
+            for character in self.nodes[id.index].data.chars() {
+                if !character.is_alphanumeric() {
+                    is_data_alphanum = false;
+                }
+            }
+            if is_data_alphanum {
+                label.push_str(&"_");
+                label.push_str(&self.nodes[id.index].data.as_str());
+            }
+        }
+        dot::LabelText::LabelStr(label.into())
+    }
+}
+
+impl<'a, T, U> dot::GraphWalk<'a, NodeId, EdgeId> for Arena<T, U> {
+    fn nodes(&self) -> dot::Nodes<'a, NodeId> {
+        let mut nodeids: Vec<NodeId> = vec![];
+        for index in 0..self.nodes.len() {
+            nodeids.push(NodeId { index: index });
+        }
+        Cow::Owned(nodeids)
+    }
+
+    fn edges(&'a self) -> dot::Edges<'a, EdgeId> {
+        Cow::Owned(self.get_edges())
+    }
+
+    fn source(&self, e: &EdgeId) -> NodeId {
+        let &(s, _) = e;
+        NodeId { index: s }
+    }
+
+    fn target(&self, e: &EdgeId) -> NodeId {
+        let &(_, t) = e;
+        NodeId { index: t }
+    }
+}
+
+// Should have same output as lrpar::parser::Node<TokId>::pretty_print().
+impl<T, U> fmt::Display for Arena<T, U>
+    where T: fmt::Display,
+          U: fmt::Display
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.nodes.is_empty() {
+            return write!(f, "");
+        }
+        let mut formatted = String::new();
+        let mut stack: Vec<NodeId> = vec![NodeId { index: 0 }];
+        while !stack.is_empty() {
+            let id = stack.pop().unwrap();
+            let node = &self.nodes[id.index];
+            if node.is_leaf() {
+                // Terminal node.
+                formatted.push_str(&format!("{}\n", node));
+            } else {
+                // Non-terminal node.
+                formatted.push_str(&format!("{}\n", node));
+                let first_child = node.first_child.unwrap();
```

Comment:
In this case I did what I saw in other libraries, I would guess that the reasoning is to have the nodes be a fixed size, whereas with a vector they will grow. It does make traversal a bit more fiddly, but soon-ish I'll add some iterators to help with that.